### PR TITLE
Refactor dnn conv functions

### LINKED ISSUE #6486: small code refactoring
Following gh-5932 and discussion with @lamblin we found a way that won't request extra tests and won't open new official interface and remove the mostly duplicate code.

- Move the functions dnn_conv, dnn_gradweight, dnn_gradinput from the file gpuarray/tests/check_dnn_conv.py to the file gpuarray/dnn.py. Rename them (what about `_dnn_conv_ex`, ...)? ex for extension? The leading underscore is not a mistake. It is to mark them as private. Make sure they don't show up in the doc. Modify the tests to use the new place of those functions.
- Modify in the file gpuarray/dnn.py the functions dnn_conv, ... to call those new private functions. This will remove the duplicate code.

### PR DESCRIPTION
**Pull request by notoraptor**
Fix #6486 

@nouiz @lamblin @abergeron

### CODE REVIEW

**Comment by notoraptor** [comment]
jenkins test this

File: theano/gpuarray/dnn.py [hunk 1]
```diff
@@ -932,6 +933,79 @@ def infer_shape(self, node, shape):
...
+    ishape = [shape_i_op(i)(img) for i in range(img.ndim)]
+    kshape = [shape_i_op(i)(kerns) for i in range(kerns.ndim)]
+    out_shp = get_conv_output_shape(ishape, kshape, desc_op.border_mode, desc_op.subsample, filter_dilation=dilation)
+    out_shp = assert_conv_shape(out_shp)
+    if beta == 0:
+        real_out = GpuAllocEmpty(dtype=img.dtype, context_name=ctx_name)(*out_shp)
+    else:
+        assert out is not None
+        out = gpu_contiguous(as_gpuarray_variable(out, ctx_name))
+        check = Assert('GpuDnnConv: qiven output (for beta not null) does not have expected shape')
```
Comment (nouiz) [inline theano/gpuarray/dnn.py]: qiven -> given

File: theano/gpuarray/dnn.py [hunk 2]
```diff
@@ -932,6 +933,79 @@ def infer_shape(self, node, shape):
...
+
+    precision = get_precision(precision, [img, topgrad], for_grad=True)
+    desc = GpuDnnConvDesc(border_mode=border_mode, subsample=subsample, dilation=dilation,
+                          conv_mode=conv_mode, precision=precision, num_groups=num_groups)(kerns_shp)
+    if beta == 0:
+        real_out = GpuAllocEmpty(dtype=img.dtype, context_name=ctx_name)(*kerns_shp)
+    else:
+        assert out is not None
+        out = gpu_contiguous(as_gpuarray_variable(out, ctx_name))
+        check = Assert('GpuDnnConvGradW: qiven output (for beta not null) does not have expected shape')
```
Comment (nouiz) [inline theano/gpuarray/dnn.py]: qiven ->given

File: theano/gpuarray/dnn.py [hunk 3]
```diff
@@ -932,6 +933,79 @@ def infer_shape(self, node, shape):
...
+
+    precision = get_precision(precision, [kerns, topgrad], for_grad=True)
+    desc = GpuDnnConvDesc(border_mode=border_mode, subsample=subsample, dilation=dilation,
+                          conv_mode=conv_mode, precision=precision, num_groups=num_groups)(kerns.shape)
+    if beta == 0:
+        real_out = GpuAllocEmpty(dtype=kerns.dtype, context_name=ctx_name)(*img_shp)
+    else:
+        assert out is not None
+        out = gpu_contiguous(as_gpuarray_variable(out, ctx_name))
+        check = Assert('GpuDnnConvGradI: qiven output (for beta not null) does not have expected shape')
```
Comment (nouiz) [inline theano/gpuarray/dnn.py]: qiven -> given

**Comment by notoraptor** [comment]
Typo fixed.
